MODAL: Add validation for letter proofing

eunjuhuss opened the pull request:
#### Description:
Added validation for letter proofing.
#### Summary:
Information from Backend, validation by letter and phone is using the same confirmation code pattern.
- Added "else if" for checking input name includes letter
- If length is not min 10 - max 12 an error message is displayed  
"Code must be 10-12 characters long, only letters and numbers allowed"
- Change variable name `phoneLengthPattern` to` shortConfirmCodeLength` for readability in code
----
- Before (without validation)
<img width="320" alt="Screenshot 2020-09-25 at 08 54 49" src="https://user-images.githubusercontent.com/44289056/94236016-cea79100-ff0c-11ea-934e-fa1e40cd72f7.png">

- After(Added validation)
<img width="314" alt="Screenshot 2020-09-25 at 08 54 21" src="https://user-images.githubusercontent.com/44289056/94236027-d23b1800-ff0c-11ea-805f-cbae8283d30d.png">

-----



#### For reviewer:
- [x] Read the above description
- [x] Reviewed the code changes
- [x] Navigate to the branch (pulled the latest version)
- [x] Run the code and been able to execute the expected function
- [x] Check any styling on both desktop and mobile sizes

njons commented on src/components/ConfirmModalForm.js:
```diff
@@ -22,12 +23,17 @@ const validate = (values, props) => {
   }
   else if(inputName.includes("phone")) {
-    if (!phoneLengthPattern.test(values.phoneConfirmDialogControl)){
+    if (!shortConfirmCodeLength.test(values.phoneConfirmDialogControl)){
+      errors[inputName] = "mobile.confirm_code_wrong_length";
+    }
+  }
+  else if(inputName.includes("letter")) {
+    if (!shortConfirmCodeLength.test(values.letterConfirmDialogControl)){
       errors[inputName] = "mobile.confirm_code_wrong_length";
```
I really like that the regex variable has been renamed from a specific "phone" name to a generic "confirm code" name!
* Is there a specific reason why the regex has `short` in its name, does it not trigger when codes are longer than 12 chars?
* Could we do something similar to the naming of `mobile.confirm_code_wrong_length` if it applies to both the phone and letter confirmation codes?

njons reviewed (commented):
👍  works really well!
- two small comments re: naming for general housekeeping

njons commented:
I have spoken to Fredrik who gave me some context for the `long`/`short` naming of the patterns and its use in the backend code, so what do you think about the following:

- Add a comment in the code to explain the naming of the patterns (please use or improve/correct the below suggestion): 
`// Naming of regex patterns refer to matching long (UUID format) and short (10 characters) codes. This aligns with naming used in the backend`
`// longCodePattern is used to verify an added email address`
`// shortCodePattern is used to verify an added phone number and vetting via letter and phone`

- Name the variables following the convention explained in the above comment

eunjuhuss commented:
It make sense to use similar descriptions in fronten and backend. I updated variable names and left comments with your suggestions.

eunjuhuss commented:
I got some suggestion from Fredrik to improve code. 
- Removed `spacePattern `and added trim function for validation
- `(!longCodePattern.test(values.emailConfirmDialogControl)` changes to `(!longCodePattern.test(trimmedCode)`/`(!shortCodePattern.test(values.phoneConfirmDialogControl)` changes to `(!shortCodePattern.test(trimmedCode)`for code readability 
- Added trim function in containers (Emails, LetterProofing, Mobile) to send the trimmed data to backend.

eunjuhuss commented:
- Confirmation code (mobile) 
![Screenshot 2020-10-05 at 11 47 45](https://user-images.githubusercontent.com/44289056/95064950-b1d73e80-0700-11eb-80fe-d54f507e0ac6.png)
- Confirmation code (By post)
![Screenshot 2020-10-05 at 11 47 52](https://user-images.githubusercontent.com/44289056/95064958-b3086b80-0700-11eb-89c9-076a1c779ecf.png)
- Email
![Screenshot 2020-10-05 at 11 48 02](https://user-images.githubusercontent.com/44289056/95064961-b4d22f00-0700-11eb-8462-b0c1fd2edaa9.png)
- Security key
![Screenshot 2020-10-05 at 11 48 09](https://user-images.githubusercontent.com/44289056/95064960-b4399880-0700-11eb-867f-9a2401c91802.png)

eunjuhuss commented:
Please see below screen shots  
First screen shot , empty value or with spaces
Second, with invalid confirmation code
Third, valid code

-----
- Confirmation code (mobile) 
![Screenshot 2020-10-05 at 11 47 45](https://user-images.githubusercontent.com/44289056/95064950-b1d73e80-0700-11eb-80fe-d54f507e0ac6.png)
- Confirmation code (By post)
![Screenshot 2020-10-05 at 11 47 52](https://user-images.githubusercontent.com/44289056/95064958-b3086b80-0700-11eb-89c9-076a1c779ecf.png)
- Email
![Screenshot 2020-10-05 at 11 48 02](https://user-images.githubusercontent.com/44289056/95064961-b4d22f00-0700-11eb-8462-b0c1fd2edaa9.png)
- Security key
![Screenshot 2020-10-05 at 11 48 09](https://user-images.githubusercontent.com/44289056/95064960-b4399880-0700-11eb-867f-9a2401c91802.png)
-----
We have two different confirmation code. one that is short code for mobile and letter, and the other one is long confirmation code for email. 
I created a regexPattern file in /login/app_utils/validation/regexPattern.js. for reusable patterns. they are imported in `Security.js` `LetterProofing.js` `Mobile.js` `Emails.js `

in the components files (`Security.js` `LetterProofing.js` `Mobile.js` `Emails.js `) send two props 
`validationPattern={shortCodePattern}` 
` validationError={"modal.short_code_invalid_format"})` 
in ConfirmModal, from ConfirmModal to ConfirmModalForm.
These two props is for rendering different error messages.

I added trim() function  in `Emails.js`, `LetterProofing.js`, `Mobile.js`, `Security.js `(Containers)for sending trimmed data to backend.

njons commented on src/login/translation/messageIndex.js:
```diff
@@ -46,15 +46,20 @@ export const unformattedMessages = defineMessages({
   },
-  "emails.invalid_code": {
-    id: "emails.invalid_code",
+  "modal.long_code_invalid_format": {
```
perhaps "modal" could be removed from all ids as we already know that we want to remove the input from modals. When we do, we would have make changes here as well, so perhaps something more generic relating to "confirmation code" could be suitable?

njons commented on src/login/translation/messageIndex.js:
```diff
@@ -46,15 +46,20 @@ export const unformattedMessages = defineMessages({
+    defaultMessage: `The code is 10 characters long`,
+    description: "Error message for mobile and letter wrong length",
+  },
+  "modal.security_code_invalid_format": {
```
Perhaps this could be renamed to something more like `"securitykey_description_invalid_format"`, to specify that it relates to security keys rather than just "security" and to the description of security keys rather than a "code".

njons commented on src/login/translation/messageIndex.js:
```diff
@@ -46,15 +46,20 @@ export const unformattedMessages = defineMessages({
-    description: "Error message for wrong length",
+  "modal.short_code_invalid_format": {
+    id: "modal.short_code_invalid_format",
+    defaultMessage: `The code is 10 characters long`,
```
I really like how you have simplified the logic for validation and the messaging! its really user friendly! 

However, I need some help understanding this one, because it tells me that the code is 10 characters, but it is 9 (see screenshot). It's also in orange so I assume something is wrong, but what I am not sure.
 
<img width="300" alt="Screenshot 2020-10-07 at 10 10 40" src="https://user-images.githubusercontent.com/30963614/95306201-60a58700-0887-11eb-8ddd-6db12e687443.png">

fredrikt replied:
I think it was me who convinced Eunju to use this phrase. I agree that it didn't turn out very well (I needed to see it in the screenshot to realise that), but the thing is that we want to convey two things in a short and simple way:

1. The code actually _is_ 10 characters long, because we made it and put it in the letter/SMS.

2. Since the currently entered code is not 10 characters long, the user probably haven't found the right code - or entered it incorrectly.

Maybe just say "Incorrect code" without trying to convey to the user that it is either the length that is clearly wrong, or an invalid character was entered?

eunjuhuss replied:
One option is to add a help text next to the label, "the code is 10 characters long" and change the error message to "incorrect code"

fredrikt replied:
I'm not sure it would actually be a helpful help text unless we start checking length separately from the format. If I enter "räksmörgås" (10 characters), "Incorrect code" is correct, but any kind of reference to the length and the number 10 would perhaps only be misleading.

njons commented on src/login/app_utils/validation/regexPattern.js:
```diff
@@ -0,0 +1,9 @@
+// Naming of regex patterns refer to matching long (UUID format) and short (10 characters) codes. This aligns with naming used in the backend
+// UUID format source from: https://en.wikipedia.org/wiki/Universally_unique_identifier#Format
+// longCodePattern is used to verify an added email address
+export const longCodePattern = /^[0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12}$/;
+// shortCodePattern is used to verify an added phone number and vetting via letter and phone
+export const shortCodePattern = /^[A-Za-z0-9]{10}$/;
+// securityKeyPattern is used to verify user's entered descriptions for security keys
+export const securityKeyPattern = /^.{1,50}$/;
+
```
Please tell me more about this file, I'm sure there is some clever idea of how we would use it.
If it contains multiple patters would it be possible to name it accordingly?

njons commented on src/components/ConfirmModalForm.js:
```diff
@@ -5,41 +5,22 @@ import { Form } from "reactstrap";
 import CustomInput from "../login/components/Inputs/CustomInput";
 
+// Check the user supplied value against a regexp pattern provided when creating the ConfirmModal.
 const validate = (values, props) => {
```
Are we still planning to move the validation functions out of the components into separate files? 
I have some reasons for why I think we should do it and (if there are any) I'd be happy to hear any reasons why they should remain in the file.

fredrikt replied:
Are you proposing to duplicate the code into each file that has a modal in it? I would say that is bad practice. You almost always end up with different bugs fixed in the different incarnations of the code.

https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

njons commented on src/components/Emails.js:
```diff
@@ -104,6 +105,8 @@ class Emails extends Component {
           handleResend={this.props.handleResend}
           handleConfirm={this.props.handleConfirm}
           helpBlock={this.props.translate("emails.confirm_help_text")}
+          validationPattern={longCodePattern}
+          validationError={"modal.long_code_invalid_format"}
```
How would this be affected by validation that has multiple possible errors to communicate to the user?

njons reviewed (commented):
Hey! Thank you so much for this work! **BIG 👍 !!** 

Thank you for going through and considering the existing code so carefully and for cleaning up so well. And thank you for refactoring the validation to make it more compact (it actually makes it more understandable). Final thanks for introducing a new way of importing the regex patterns for validation. 

I have some questions, mainly because I am excited to learn more about your solutions and how they will make our code better in the long run. And some very minor thoughts on some message id names completely inspired by your level of thoughtfulness and care!

eunjuhuss commented:
- Remove `confirmation.short.code_invalid_format` and Rename `confirmation.long.code_invalid_format `to `confirmation.code_invalid_format` for reusability
- Rename file `regexPattern` to `regexPatterns`
- Rename error message Id `modal.security_code_invalid_format` to `security.description_invalid_format`

fredrikt reviewed (commented):
LGTM
